Export licence data

leenagupte opened the pull request:
Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

Trello: https://trello.com/c/DY5YCALU

# What's changed?
Adds a JSON endpoint to export the following data for each licence:

- gds_id
- Devolved administrations
- Level 2 sectors

# Why?

Once the data has been exported, it will be imported into Specialist Publisher and used to tag licences to industry sectors in the new "Find Licences" specialist finder.

Facets in specialist finders do not support nested facets, therefore only one level of sectors was required, and level two was selected by the user researchers and content designers for the first round of testing.

Activities were are not being exported as they have an almost one to one mapping with licences, but the activity names do not match the licence names.

1pretz1 commented on app/lib/data_exporter.rb:
```diff
@@ -0,0 +1,42 @@
+    sectors = Sector.in(id: LicenceLink.where(licence_id: licence.id).pluck(:sector_id))
+    sectors.map do |sector|
+      parent_sectors = sector.parents.to_a
+      parent_sectors.first.name.parameterize if parent_sectors.first&.layer == 2
```
We should probably record and print out when / if this happens to see if our assumption is correct. Currently we'll just get `nil` values in the array if this is true.

leenagupte replied:
Hmm that's true. I think I might have fallen foul of this when I was setting up the tests. I'll remove the line and test it against integration again.

leenagupte replied:
Actually, can I clarify? 

Do you think the guard clause `if parent_sectors.first&.layer == 2` isn't needed at all, or just the safe navigation operator?

leenagupte replied:
Ok. I'll have a little play around in integration out of curiosity,  however I don't think the guard clause hurts in general

leenagupte replied:
Ok, so there aren't any LicenceLinks to layer 1 or 2 sectors on integration: https://deploy.integration.publishing.service.gov.uk/job/run-rake-task/305278/console

Some of the licences are tagged to an astonishing amount of layer 3 sectors though. I think I saw 89 for one of them!

1pretz1 replied:
Thanks for checking, wow... 89 sectos 😱

If we've confirmed there are none, then we can remove the guard clause completely as the data isn't going to change 👍

leenagupte replied:
Oh, I should mention, that I've replaced `map` with `filter_map` as that's the equivalent of doing `map` followed by `compact` to remove nils.

1pretz1 replied:
Okay, sure (concerning guard clause) 👍  

> Oh, I should mention, that I've replaced map with filter_map as that's the equivalent of doing map followed by compact to remove nils.

Ah nice, I didn't know `filter_map` existed!

1pretz1 commented on app/lib/data_exporter.rb:
```diff
@@ -0,0 +1,42 @@
+  end
+
+  def call
+    Rails.logger.info licence_data.each(&:deep_stringify_keys!).to_yaml
```
It be nice to have this output tested but not a blocker as this will only be a temporary task. You've commented the results in a commit which feels like a good half way house 👍

leenagupte replied:
Yeah, I didn't see a need to test `to_yaml` when we are already testing the output from a nice readable hash.

1pretz1 reviewed (commented):
Nice work Leena! Only one small comment. 

Your approach also makes sense of exporting all the licences (even the ones that don't exist in Publisher) as we have the list of common_licence_identifiers in SP which we can use or we can just call Publishing API to return all the licence_transactions and match accordingly.

leenagupte commented:
>Your approach also makes sense of exporting all the licences (even the ones that don't exist in Publisher) as we have the list of common_licence_identifiers in SP which we can use or we can just call Publishing API to return all the licence_transactions and match accordingly.

@1pretz1 My though was to just update your existing import script to tag the licences as well by matching on licence identifier.

leenagupte commented:
@1pretz1 Thanks for your comments. I've updated the PR to remove duplicate level 2 sectors. That has taken the length of the yaml output from 10,473 lines to 4,476 lines 😱 

The rake task has been retested in integration and the console output and the gist linked in the PR description have been updated with the new results.

This is now ready for a re-review.

leenagupte commented:
@1pretz1 Considering that there are so many lines, I'm wondering now if this should instead be an API endpoint.

1pretz1 commented:
Changes are looking good @leenagupte! Just one inline comment and some general points below:

> My though was to just update your existing import script to tag the licences as well by matching on licence identifier.

That's a good idea 👍 I'd just be mindful of the class size / complexity as we may need to split the importer class up as it grows.

> Considering that there are so many lines, I'm wondering now if this should instead be an API endpoint.

It could be an API endpoint and if we were certain that it is going to be reused then it probably should be, however as there is more setup involved and this may be a one off task, it seems like we don't gain a huge deal (setup vs outcome) by exposing a new API endpoint - what do you think?  

We could also make the task write directly to a file which is a bit easier to work with vs copying and pasting, as you can scp onto the machine and transfer the file across. However, as you've already saved the results, this comment is less relevant 👍 

We also have a few errors from the job file, assume these are unrelated?

```
14:56:43 /usr/lib/rbenv/versions/3.1.2/lib/ruby/3.1.0/unicode_normalize/normalize.rb:21: warning: Exception in finalizer #<Proc:0x00007fa734747d80 /data/vhost/licencefinder/shared/bundle/ruby/3.1.0/gems/mongo-2.14.0/lib/mongo/cursor.rb:108>
14:56:43 /data/vhost/licencefinder/shared/bundle/ruby/3.1.0/gems/mongo-2.14.0/lib/mongo/cluster/reapers/cursor_reaper.rb:60:in `synchronize': can't be called from trap context (ThreadError)
```

leenagupte commented:
> It could be an API endpoint and if we were certain that it is going to be reused then it probably should be, however as there is more setup involved and this may be a one off task, it seems like we don't gain a huge deal (setup vs outcome) by exposing a new API endpoint - what do you think?

I've added a [commit](https://github.com/alphagov/licence-finder/pull/1275/commits/8b505cde717731af497104c7874be0b3bb7dc9ef) with the changes to make an API endpoint, and they really are very small. Having an api endpoint means that we can run the import scripts in Specialist Publisher multiple times in integration for testing without any extra hassle.

It will be easy to fix these changes up.

> 
> We could also make the task write directly to a file which is a bit easier to work with vs copying and pasting, as you can scp onto the machine and transfer the file across. However, as you've already saved the results, this comment is less relevant 👍
> 

I want to avoid copying and pasting as I found it's not easy to select 4,747 lines from the Jenkins log. Also having to SCP is another hassle for the developer.

I also had warnings in my editor about the size of the YAML file.

I do think it would be much easier to render the JSON api and read in the data in Specialist Publisher. That way we can always point at production data, and it will take very little effort to run the import job as and when needed.

> We also have a few errors from the job file, assume these are unrelated?
> 
> ```
> 14:56:43 /usr/lib/rbenv/versions/3.1.2/lib/ruby/3.1.0/unicode_normalize/normalize.rb:21: warning: Exception in finalizer #<Proc:0x00007fa734747d80 /data/vhost/licencefinder/shared/bundle/ruby/3.1.0/gems/mongo-2.14.0/lib/mongo/cursor.rb:108>
> 14:56:43 /data/vhost/licencefinder/shared/bundle/ruby/3.1.0/gems/mongo-2.14.0/lib/mongo/cluster/reapers/cursor_reaper.rb:60:in `synchronize': can't be called from trap context (ThreadError)
> ```

Hm I haven't seen these errors before. Locally I get a different set of errors. But Jenkins and GitHub Actions are happy 😱

1pretz1 commented:
> I do think it would be much easier to render the JSON api and read in the data in Specialist Publisher. That way we can always point at production data, and it will take very little effort to run the import job as and when needed.

Ah @leenagupte, when I made the comment about API, I was envisioning a private API which would require you to setup tokens etc. Yes a publicly available API would be much simpler / easier to work with and as we're only exposing any data already in the public domain then this seems a good solution 👍

leenagupte commented:
> Ah @leenagupte, when I made the comment about API, I was envisioning a private API which would require you to setup tokens etc. Yes a publicly available API would be much simpler / easier to work with and as we're only exposing data already in the public domain then this seems a good solution 👍

Oh yeah, a private API would be a major hassle. But as you said the data being exposed is already public either on each licence page, or via search data. I've also set it not to index the api in search as with the `/browse-licences` endpoint.

@1pretz1 I've fixed up the changes to add an API and removed the rake task. This PR is now a bit smaller and ready for review.

1pretz1 reviewed (approved):
PR looks good to me, nice job! Using an API is a much nicer approach 🎉 🚢
